godot: 3.2.2 -> 3.3.2

robwhitaker opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Update to the most recent version of Godot 3.x.

###### Additional Notes

I'm not quite sure why the patch files are necessary (or if they still are), but I've updated them to cleanly apply to the updated Godot codebase.

Additionally, I had to add an extra build option (`udev = false;`) to get it to build. If I read [Godot's build docs](https://docs.godotengine.org/en/stable/development/compiling/compiling_for_x11.html#doc-compiling-for-x11) correctly, the default is supposed to be `false`, but without explicitly adding the flag, it tries to build with `libudev`.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

robwhitaker committed:
godot: 3.2.2 -> 3.3

yusdacra commented on pkgs/development/tools/godot/default.nix:
```diff
@@ -7,16 +7,17 @@ let
   options = {
     touch = libXi != null;
     pulseaudio = false;
+    udev = false;
```
Perhaps we could do
```suggestion
    udev = libudev != null;
```
like the `touch` option?

robwhitaker replied:
That's the more ideal way to do it, but it doesn't quite work. Even with the `libudev` library included in `buildInputs`, the build process throws:

```
platform/x11/libudev-so_wrap.c:96:10: fatal error: libudev.h: No such file or directory
   96 | #include <libudev.h>
```

So just including `libudev` doesn't allow us to actually enable it for the build. I'd imagine the fix for this is an extra line of wiring somewhere, but I don't really know where. Maybe patching in another `env.ParseConfig` (something like `env.ParseConfig("pkg-config libudev --cflags --libs")`) would do it, but that's mostly an uninformed hunch.

robwhitaker replied:
Looks like my hunch was right. Should build with or without `libudev` now.

robwhitaker committed:
Build Godot with libudev if it is available

yusdacra commented:
Result of `nixpkgs-review pr 123126` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>godot</li>
    <li>godot-export-templates</li>
    <li>godot-headless</li>
    <li>godot-server</li>
    <li>oh-my-git</li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/development/tools/godot/default.nix:
```diff
@@ -1,27 +1,28 @@
-{ stdenv, lib, fetchFromGitHub, scons, pkg-config, libX11, libXcursor
-, libXinerama, libXrandr, libXrender, libpulseaudio ? null
+{ stdenv, lib, fetchFromGitHub, scons, pkg-config, libudev ? null, libX11
+, libXcursor , libXinerama, libXrandr, libXrender, libpulseaudio ? null
 , libXi ? null, libXext, libXfixes, freetype, openssl
 , alsaLib, libGLU, zlib, yasm ? null }:
```
```suggestion
{ stdenv, lib, fetchFromGitHub, scons, pkg-config, libudev ? null, libX11
, libXcursor , libXinerama, libXrandr, libXrender, libpulseaudio
, libXi, libXext, libXfixes, freetype, openssl
, alsaLib, libGLU, zlib, yasm }:
```

If those packages are null you are doing something wrong here.

robwhitaker replied:
Why is that? Am I right in assuming it is because the top level call to `callPackage` should always fill those dependencies in, so they should never be `null`?

In the case of `libudev` and `yasm`, these packages should be optional to the build. Is there a common convention for handling this sort of thing?

SuperSandro2000 replied:
> Why is that? Am I right in assuming it is because the top level call to callPackage should always fill those dependencies in, so they should never be null?

Yes, they can only be null when you overwrite them to null.

> In the case of libudev and yasm, these packages should be optional to the build. Is there a common convention for handling this sort of thing?

It would be easier to discover if there was a boolean flag like enableUdev or withUdev.

robwhitaker replied:
When it comes to boolean flags, is there a preference between defaulting to false (opt-in) or defaulting to true (opt-out)? I'd probably go for the former, but is there a preferred approach within nixpkgs?

yusdacra replied:
> When it comes to boolean flags, is there a preference between defaulting to false (opt-in) or defaulting to true (opt-out)? I'd probably go for the former, but is there a preferred approach within nixpkgs?

I think we should keep the current behaviour, which should be defaulting to true I believe.

robwhitaker replied:
Well, I removed the null defaults and updated to 3.3.2, but now the tests seem to be failing. Maybe I'm reading it wrong, but:

`pkgs/development/tools/godot/default.nix:1:1 called without required argument 'libudev'`

seems to indicate that it *is* indeed possible for `libudev` to not exist (making the `null` default necessary). Am I reading this right or is it something else?

yusdacra replied:
Then we can default `withUdev` to `false`. From what I see in the latest commit, `libudev` is always put in `buildInputs`, this should be done like `buildInputs = [ ... ] ++ (lib.optional options.withUdev libudev)`, so that it's not needed if `withUdev` is false. Same could be done for other options too.

robwhitaker commented:
At this point, Godot has had two more patch releases (3.3.1 and 3.3.2). Does it make more sense to update this PR to 3.3.2 or finish this one as-is and create a new PR for the newer Godot version?

yusdacra commented:
> At this point, Godot has had two more patch releases (3.3.1 and 3.3.2). Does it make more sense to update this PR to 3.3.2 or finish this one as-is and create a new PR for the newer Godot version?

I think updating this PR to 3.3.2 should be fine since this is already a package update.

robwhitaker committed:
Remove null defaults from Godot package params and add withUdev flag

* Since these packages are coming from a `callPackage` call, they should
  always exist, making the `null` defaults redundant.
* Using `withUdev` instead of a null check makes it easier to discover
  the option and clearer what it is doing (building with udev support).

robwhitaker committed:
godot: 3.3 -> 3.3.2

SuperSandro2000 commented on pkgs/development/tools/godot/default.nix:
```diff
@@ -1,27 +1,30 @@
-, libXinerama, libXrandr, libXrender, libpulseaudio ? null
-, libXi ? null, libXext, libXfixes, freetype, openssl
-, alsaLib, libGLU, zlib, yasm ? null }:
+{ stdenv, lib, fetchFromGitHub, scons, pkg-config, libudev, libX11
```
```suggestion
{ stdenv, lib, fetchFromGitHub, scons, pkg-config, udev, libX11
```

SuperSandro2000 committed:
Update pkgs/development/tools/godot/default.nix

SuperSandro2000 commented on pkgs/development/tools/godot/default.nix:
```diff
@@ -1,27 +1,30 @@
   nativeBuildInputs = [ pkg-config ];
   buildInputs = [
-    scons libX11 libXcursor libXinerama libXrandr libXrender
+    scons libudev libX11 libXcursor libXinerama libXrandr libXrender
```
```suggestion
    scons udev libX11 libXcursor libXinerama libXrandr libXrender
```

SuperSandro2000 committed:
Update pkgs/development/tools/godot/default.nix
